File: api/models/segments.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Pydantic models for segments API
Author: Customer Data Analytics Team
"""

from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import date

class SegmentDistribution(BaseModel):
    """Модель распределения сегмента"""
    cluster_id: int = Field(..., description="ID кластера")
    users_count: int = Field(..., description="Количество пользователей")
    share: float = Field(..., description="Доля от общего числа пользователей")

class SegmentsDistributionResponse(BaseModel):
    """Ответ API для распределения сегментов"""
    date: str = Field(..., description="Дата снапшота")
    timezone: str = Field(..., description="Часовой пояс")
    available: bool = Field(..., description="Доступность данных")
    total_users: int = Field(..., description="Общее количество пользователей")
    segments: List[SegmentDistribution] = Field(..., description="Список сегментов")

class SegmentPoint(BaseModel):
    """Точка данных для динамики сегмента"""
    date: str = Field(..., description="Дата")
    count: int = Field(..., description="Количество пользователей")

class SegmentSeries(BaseModel):
    """Серия данных для кластера"""
    cluster_id: int = Field(..., description="ID кластера")
    points: List[SegmentPoint] = Field(..., description="Точки данных")

class SegmentsDynamicsResponse(BaseModel):
    """Ответ API для динамики сегментов"""
    from_date: str = Field(..., alias="from", description="Начальная дата")
    to_date: str = Field(..., alias="to", description="Конечная дата")
    timezone: str = Field(..., description="Часовой пояс")
    series: List[SegmentSeries] = Field(..., description="Серии данных по кластерам")

class MigrationMatrix(BaseModel):
    """Элемент матрицы переходов"""
    from_cluster: int = Field(..., alias="from", description="Исходный кластер")
    to_cluster: int = Field(..., alias="to", description="Целевой кластер")
    count: int = Field(..., description="Количество переходов")

class SegmentsMigrationResponse(BaseModel):
    """Ответ API для матрицы переходов"""
    date: str = Field(..., description="Дата анализа")
    prev_date: str = Field(..., description="Предыдущая дата")
    timezone: str = Field(..., description="Часовой пояс")
    available: bool = Field(..., description="Доступность данных")
    matrix: List[MigrationMatrix] = Field(..., description="Матрица переходов")
    note: Optional[str] = Field(None, description="Дополнительная информация")

class ClusterMeta(BaseModel):
    """Метаданные кластера"""
    id: int = Field(..., description="ID кластера")
    name: str = Field(..., description="Название кластера")
    description: str = Field(..., description="Описание кластера")

class SegmentsMetaResponse(BaseModel):
    """Ответ API для метаданных сегментов"""
    clusters: List[ClusterMeta] = Field(..., description="Список кластеров")

File: customer-data-analytics/frontend/src/services/segments.ts
import { httpGet } from './http'

export type ClusterMeta = { id: number; name: string; description?: string }

export type DistributionItem = { cluster_id: number; users_count: number; share: number }
export type DistributionResponse = {
  date: string
  timezone: string
  available: boolean
  total_users: number
  segments: DistributionItem[]
  last_available_date?: string
  note?: string
}

export type DynamicsPoint = { date: string; count: number }
export type DynamicsSeries = { cluster_id: number; points: DynamicsPoint[] }
export type DynamicsResponse = {
  from_date: string
  to_date: string
  granularity: 'day' | 'week' | 'month'
  timezone: string
  available: boolean
  series: DynamicsSeries[]
  note?: string
}

export type MigrationCell = { from_cluster: number; to_cluster: number; count: number }
export type MigrationResponse = {
  date: string
  timezone: string
  available: boolean
  matrix: MigrationCell[]
  note?: string
}

const API_BASE = '/api/v1/segments'

export function fetchMeta() {
  return httpGet<ClusterMeta[]>(`${API_BASE}/meta`)
}

export function fetchDistribution(params?: { date?: string }) {
  return httpGet<DistributionResponse>(`${API_BASE}/distribution`, params)
}

export function fetchDynamics(params: { from_date: string; to_date: string; granularity: 'day' | 'week' | 'month' }) {
  return httpGet<DynamicsResponse>(`${API_BASE}/dynamics`, params)
}

export function fetchMigration(params: { date: string }) {
  return httpGet<MigrationResponse>(`${API_BASE}/migration`, params)
}



